harmony-one/harmony
# Staged Stream Sync v1.0

## Issue

This pull request adds the initial version of **Staged Stream Sync**, which is a new feature for syncing in sequential stages. The idea of staged sync is given from Erigon, which could improve the performance of full syncing in Ethereum.

Each stage is executed in order, and a stage N does not stop until the local head reaches for it. The current implemented staged stream sync has 6 stages and a couple of new stages could be added later.

Current stages are:

- **Heads**: Retrieve Chain Heads

- **ShortRange**: Short Range Sync, it executes only if node is synced.

- **SyncEpoch**:  Epoch Sync

- **BlockBodies**: Download block bodies

- **States**: Insert new blocks and update blockchain states.

- **Finish**: Final stage to finalize the process. (will be used in next phases to create indices)

## Features:

- No need any more for DNS nodes. Nodes talk to each other and share the states.

- Faster full sync compared to current legacy syncing

- More stable against network issues, because a few stages can be run offline

- 100% backwards compatible

- Can be used in the main net without any changes to the current main net structure.

- Efficiently using of CPU and RAM and Bandwidth 

- Using Erigon DB 

- Using single/multiple transactions to read/write date from/to db

- Supporting of memory and disc DB, based on node specs

- Can resume syncing after any crash

- Fully configurable to sync in a single cycle or multiple cycles

- Improvements in resolving bad blocks


## Test

- all of the shard/beacon nodes has to have running stream server. (Sync server has to be true)
- make sure StagedSync and Downloader is set to true in config file
- run a node and wait for it to be fully synced

****
 
### Test/Run Logs

## Operational Checklist

1. **Does this PR introduce backward-incompatible changes to the on-disk data structure and/or the over-the-wire protocol?**. (If no, skip to question 8.)

    **NO**

2. **Describe the migration plan.**. For each flag epoch, describe what changes take place at the flag epoch, the anticipated interactions between upgraded/non-upgraded nodes, and any special operational considerations for the migration.

Staged sync is fully backward compatible and there is no need to touch anything in the main net. Any node that uses the new version will be able to use it. It only needs to enable staged stream sync for the nodes.

3. **Describe how the plan was tested.**

Tested simply by running a node and monitoring syncing for a couple of hours.

4. **Does this PR introduce backward-incompatible changes *NOT* related to on-disk data structure and/or over-the-wire protocol?** (If no, continue to question 11.)

    **NO**

5. **Does this PR introduce significant changes to the operational requirements of the node software, such as a >20% increase in CPU, memory, and/or disc usage?**

It will use the CPU and bandwidth more efficiently.
